W3C.Domain/Repositories/MongoDbTransactionCoordinator.cs

```diff
@@ -0,0 +1,103 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Bson;
+using MongoDB.Driver;
+using Serilog;
+
+namespace W3C.Domain.Repositories;
+
+public class MongoDbTransactionCoordinator : ITransactionCoordinator
+{
+    private readonly IMongoClient _mongoClient;
+    private IClientSessionHandle _session;
+    private ConcurrentQueue<Func<Task>> _onSuccessHandlers;
+
+    public bool IsTransactionActive => _session != null && _session.IsInTransaction;
+
+    public MongoDbTransactionCoordinator(IMongoClient mongoClient)
+    {
+        _mongoClient = mongoClient ?? throw new ArgumentNullException(nameof(mongoClient));
+        _onSuccessHandlers = new ConcurrentQueue<Func<Task>>();
+    }
+
+    public async Task BeginTransactionAsync(CancellationToken cancellationToken = default)
+    {
+        if (IsTransactionActive)
```

Comment:
Does it make sense that we can only ever have a single request happening at once per collection, otherwise we throw errors? Would it not be better to instead have multiple sessions and let MongoDB deal with ordering transactions as needed? 

It seems to me that throwing exceptions whenever there's more than 1 request at a time is probably not ideal.

Reply:
Oh boy, you're right! At some point I had the session variable within an AsyncLocal but somehow the contexts got messed up and I removed it again. That's why I had this who IsTransactionActive be a dynamic call. Yeah, that's a very good call, thank you!